go/saltpack/sign_stream.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package saltpack
+
+import (
+	"bytes"
+	"io"
+)
+
+type signAttachedStream struct {
+	header      *SignatureHeader
+	wroteHeader bool
+	encoder     encoder
+	buffer      bytes.Buffer
+	block       []byte
+	seqno       PacketSeqno
+	secretKey   SigningSecretKey
+}
+
+func newSignAttachedStream(w io.Writer, signer SigningSecretKey) (*signAttachedStream, error) {
+	if signer == nil {
+		return nil, ErrInvalidParameter{message: "no signing key provided"}
+	}
+
+	header, err := newSignatureHeader(signer.PublicKey(), MessageTypeAttachedSignature)
+	if err != nil {
+		return nil, err
+	}
+
+	stream := &signAttachedStream{
+		header:    header,
+		encoder:   newEncoder(w),
+		block:     make([]byte, SignatureBlockSize),
+		secretKey: signer,
+	}
+
+	return stream, nil
+}
+
+func (s *signAttachedStream) Write(p []byte) (int, error) {
+	if !s.wroteHeader {
+		s.wroteHeader = true
```

Comment:
Ok, but it will also be weird if they ignore the error and get in an infinite loop failing to write header blocks.